File: courses/codes/GraphSearchSample.java

import java.util.*;

public class GraphSearchSample {

    // Creates the adjacency set representation of a sample graph
    public static Map<Integer, Set<Integer>> sampleGraph() {
        Map<Integer, Set<Integer>> graph = new HashMap<>();

        graph.put(0, new HashSet<>(Arrays.asList(1, 2)));
        graph.put(1, new HashSet<>(Arrays.asList(1, 2)));
        graph.put(2, new HashSet<>(Arrays.asList(0, 1, 3, 4)));
        graph.put(3, new HashSet<>(Arrays.asList(2)));
        graph.put(4, new HashSet<>(Arrays.asList(2)));

        return graph;
    }

    // Returns the node set N given the adjacency list G
    public static Set<Integer> getNodeSet(Map<Integer, Set<Integer>> G) {
        return G.keySet();
    }

    // Returns the edge set E given the adjacency list G
    public static Set<String> getEdgeSet(Map<Integer, Set<Integer>> G, boolean directed) {
        Set<String> edges = new HashSet<>();
        for (Map.Entry<Integer, Set<Integer>> entry : G.entrySet()) {
            int from = entry.getKey();
            for (int to : entry.getValue()) {
                if (directed || from < to) {
                    edges.add(from + "," + to);
                }
            }
        }
        return edges;
    }

    // Apply the BFS Algorithm on graph G
    public static void bfs(Map<Integer, Set<Integer>> G) {
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> nodes = getNodeSet(G);
        Map<Integer, Boolean> visited = new HashMap<>();
        Map<Integer, Integer> level = new HashMap<>();

        for (int node : nodes) {
            visited.put(node, false);
            level.put(node, Integer.MAX_VALUE);
        }

        int source = 0;
        queue.add(source);
        visited.put(source, true);
        level.put(source, 0);

        while (!queue.isEmpty()) {
            int node = queue.poll();
            System.out.println("Level {" + level.get(node) + "}: {" + node + "}");

            for (int neighbor : G.get(node)) {
                if (!visited.get(neighbor)) {
                    visited.put(neighbor, true);
                    queue.add(neighbor);
                    level.put(neighbor, level.get(node) + 1);
                }
            }
        }
    }

    // Apply the DFS Algorithm on graph G
    public static void dfs(Map<Integer, Set<Integer>> G) {
        Stack<Integer> stack = new Stack<>();
        Set<Integer> nodes = getNodeSet(G);
        Map<Integer, Boolean> visited = new HashMap<>();

        for (int node : nodes) {
            visited.put(node, false);
        }

        int source = 0;
        stack.push(source);

        while (!stack.isEmpty()) {
            int node = stack.pop();
            if (!visited.get(node)) {
                visited.put(node, true);
                System.out.println("Visited node {" + node + "}");
                for (int neighbor : G.get(node)) {
                    if (!visited.get(neighbor)) {
                        stack.push(neighbor);
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        Map<Integer, Set<Integer>> G = sampleGraph();

        System.out.println("BFS: ");
        bfs(G);

        System.out.println("DFS: ");
        dfs(G);
    }
}

File: courses/codes/cclv1-sr-class/graph_search_sample.py
'''
    Creates the adjacency list/set representation of a sample graph
'''
def sample_graph():
    return {
        0: {1, 2},
        1: {1, 2},
        2: {0, 1, 3, 4},
        3: {2},
        4: {2}
    }

'''
    Returns the node set N given adj list G
'''
def get_node_set(G):
    return set(G.keys())

'''
    Returns the edge set E given adj list G
'''
def get_edge_set(G, directed=False):
    edges = set()
    for k, v in G:
        for node in v:
            if directed or k < node:
                edges.add((k, node))
    return edges

'''
    Apply the BFS Algorithm on graph G
'''
def bfs(G):
    Q = []                          # initialize queue as list
    N = get_node_set(G)             # initialize nodes
    V = [False] * len(N)            # initialize boolean array; must be boolean dict/map for arbitrary nodes
    B = [float('inf')] * len(N)     # initialize inf breadth levels

    node = 0                        # let zero be the starting node
    Q.append(node)                  # enqueue source node
    V[node] = True                  # mark source node as visited
    B[node] = 0                     # source node has breadth 0

    while len(Q) > 0:

        node = Q.pop(0)                         # dequeue from Q
        print('Level {%d}: {%d}' % (B[node], node))        # print out the node

        for nbr in G[node]:                     # For each neighbor node
            if not V[nbr]:                      # Do the following if not yet visited
                V[nbr] = True                   # mark as visited
                Q.append(nbr)                   # add that to the queue
                B[nbr] = B[node] + 1            # increment its breadth

'''
    Apply the DFS Algorithm on graph G
'''
def dfs(G):
    S = []                          # initialize stack as list
    N = get_node_set(G)             # initialize nodes
    V = [False] * len(N)            # initialize boolean array; must be boolean dict/map for arbitrary nodes

    node = 0                        # let zero be the starting node
    S.append(node)                  # push the initial source node to the stack

    while len(S) > 0:

        node = S.pop()              # pop from the last of the stack
        if not V[node]:
            V[node] = True                          # mark node visited
            print('Visited node {%d}' % node)       # print node
            for nbr in G[node]:                     # For each neighbor node do as folows:
                if not V[nbr]:
                    S.append(nbr)                   # if not yet visited, push to the stack




G = sample_graph()

print('BFS: ')
bfs(G)

print('DFS: ')
dfs(G)